pkg/controller/v1alpha1/inferencegraph/controller.go

```diff
@@ -0,0 +1,266 @@
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/go-logr/logr"
+	"github.com/golang/protobuf/proto"
+	v1alpha1api "github.com/kserve/kserve/pkg/apis/serving/v1alpha1"
+	"github.com/kserve/kserve/pkg/constants"
+	"github.com/kserve/kserve/pkg/utils"
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierr "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/retry"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/kmp"
+	"knative.dev/serving/pkg/apis/autoscaling"
+	knservingv1 "knative.dev/serving/pkg/apis/serving/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// InferenceGraphReconciler reconciles a InferenceGraph object
+type InferenceGraphReconciler struct {
+	client.Client
+	Log      logr.Logger
+	Scheme   *runtime.Scheme
+	Recorder record.EventRecorder
+}
+
+// InferenceGraphState describes the Readiness of the InferenceGraph
+type InferenceGraphState string
+
+const (
+	InferenceGraphNotReadyState InferenceGraphState = "InferenceGraphNotReady"
+	InferenceGraphReadyState    InferenceGraphState = "InferenceGraphReady"
+)
+
+type RouterConfig struct {
+	Image         string `json:"image"`
+	CpuRequest    string `json:"cpuRequest"`
+	CpuLimit      string `json:"cpuLimit"`
+	MemoryRequest string `json:"memoryRequest"`
+	MemoryLimit   string `json:"memoryLimit"`
+}
+
+func getRouterConfigs(configMap *v1.ConfigMap) (*RouterConfig, error) {
+
+	routerConfig := &RouterConfig{}
+	if agentConfigValue, ok := configMap.Data["router"]; ok {
+		err := json.Unmarshal([]byte(agentConfigValue), &routerConfig)
+		if err != nil {
+			panic(fmt.Errorf("Unable to unmarshall agent json string due to %v ", err))
+		}
+	}
+
+	//Ensure that we set proper values for CPU/Memory Limit/Request
+	resourceDefaults := []string{routerConfig.MemoryRequest,
+		routerConfig.MemoryLimit,
+		routerConfig.CpuRequest,
+		routerConfig.CpuLimit}
+	for _, key := range resourceDefaults {
+		_, err := resource.ParseQuantity(key)
+		if err != nil {
+			return routerConfig, fmt.Errorf("Failed to parse resource configuration for router: %q",
+				err.Error())
+		}
+	}
+
+	return routerConfig, nil
+}
+
+func (r *InferenceGraphReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = context.Background()
+
+	// Fetch the InferenceService instance
+	graph := &v1alpha1api.InferenceGraph{}
+	if err := r.Get(ctx, req.NamespacedName, graph); err != nil {
+		if apierr.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
```

Comment:
Question:
Is there any validation on the graph configuration? I mean should the service url being existing inferenceservice endpoint or just any string is accepted? 
Also I saw in the router image there's no validation too so it's just behave like proxy and user need make the request following our request spec? 
In fact if there's no any validation user may chain any knative service and make calls as long as the request is valid for the service accordingly. That may be an additional flexibility.

Reply:
The idea is to support both `InferenceService` name and url, if `InferenceService` name is provided then the controller fills in the url from the status object. Will also add the graph validation and graph status based on the individual inference service status.

Graph orchestrator serves as the proxy and  I think the request validation should be done in the destination service and the proxy can propagate the error back.

Reply:
That's clear to me now, thanks for response and look forward to the feature.

Reply:
if `InferenceService` name is provided we resolve to the service url and in the router it is always use service url to invoke the request.